src/t_hash.c

```diff
@@ -515,29 +515,19 @@ robj *hashTypeDup(robj *o) {
 
     serverAssert(o->type == OBJ_HASH);
 
-    switch (o->encoding) {
-        case OBJ_ENCODING_ZIPLIST:
-            hobj = createHashObject();
-            break;
-        case OBJ_ENCODING_HT:
-            hobj = createHashObject();
-            hashTypeConvert(hobj, OBJ_ENCODING_HT);
-            dict *d = o->ptr;
-            dictExpand(hobj->ptr, dictSize(d));
-            break;
-        default:
-            serverPanic("Wrong encoding.");
-            break;
-    }
-
     if(o->encoding == OBJ_ENCODING_ZIPLIST){
         unsigned char *zl = o->ptr;
         size_t sz = ziplistBlobLen(zl);
         unsigned char *new_zl = zmalloc(sz);
         memcpy(new_zl, zl, sz);
-        zfree(hobj->ptr);
-        hobj->ptr = new_zl;
+        hobj = createObject(OBJ_HASH, new_zl);
+        hobj->encoding = OBJ_ENCODING_ZIPLIST;
     } else if(o->encoding == OBJ_ENCODING_HT){
+        hobj = createHashObject();
+        hashTypeConvert(hobj, OBJ_ENCODING_HT);
```

Comment:
in the spirit of this PR, maybe you should create an robj with a dict here too, rather than create it with a ziplist and then convert it to dict.

Reply:
Thanks for suggestion, I've made changes.